Add support for local debugging in Unity and C# SDKs

nima-ap-zz opened the pull request:
Complete implementation of SDK changes necessary to support local-debugging for the Unity and C# SDK. 
This change comes in various places like adding properties to some models, changing `getRequestActions` in `make.js` to account for conditions necessary on `ExecuteFunction` API call to differentiate between local and remote calls, and other places.
Furthermore, a few other templates and utility classes have been modified to support this functionality.

nima-ap committed:
Add support for local debugging in Unity and C# SDKs

MGudgin commented on targets/csharp/make.js:
```diff
@@ -279,33 +288,34 @@ function getAuthParams(apiCall, isApiInstance = false) {
 }
 
 function getRequestActions(tabbing, apiCall, isApiInstance = false) {
```
This function doesn't look right to me. In the GetEntityToken case you will emit both the block covered by the if at line 291/301 *and* the block covered by the if at line 307/317. At best there is duplicate code. At worst it won't compile.

You should either put the return statements back and duplicate the block covered by the if at line 285/295 in the new ExecuteFunction block (and probably move the if for ExecuteFunction up about the ones for apiCall.Auth) or remove the code in the if block for GetEntityToken in line 309/319-316/326. The latter seems riskier as the code that then gets generated for GetEntityToken won't be quite the same as it is before this change.

Note: It's not actually clear to me how the current GetEntityToken block ever gets emitted. Perhaps apiCall.auth has some other value for GetEntityToken? #Closed

MGudgin replied:
OK, I'm going to assume that for GetEntityToken apiCall.auth has some value not covered by the if cases, in which case, I think this code does work for GetEntityToken. But I think you could still emit bad code for the LoginResult/RegisterPlayFabUserRequest case. I think it is safer to just keep return statements and have each if block emit all the code needed for that case. #Closed

nima-ap-zz replied:
As per our conversation, I will be making another function purely for custom api behavior with `ExecuteFunction`'s custom behavior being the first in there to avoid the fact that if we go back to return statements then `apiCall.auth === "EntityToken"` would get resolved before `apiCall.name === "ExecuteFunction"` or vice versa. #Closed

MGudgin commented on targets/csharp/source/source/PlayFabHttp/PlayFabHttp.cs:
```diff
@@ -49,6 +49,16 @@ public static class PlayFabHttp
+            return await _DoPost(fullPath, request, authType, authKey, extraHeaders);
+        }
+
+        public static async Task<object> DebugDoPost(string fullUri, PlayFabRequestCommon request, string authType, string authKey, Dictionary<string, string> extraHeaders, PlayFabApiSettings apiSettings = null)
```
Can we call this DoPostDebug or DoPostWithFullUri? #Closed

nima-ap-zz replied:
DoPostWithFullUri sounds good to me! #Closed

MGudgin commented on targets/csharp/templates/API.cs.ejs:
```diff
@@ -17,6 +17,7 @@ namespace PlayFab
         {
 <%- getRequestActions("            ", apiCall) %>
             var httpResult = await PlayFabHttp.DoPost("<%- apiCall.url %>", request, <%- getAuthParams(apiCall) %>, extraHeaders);
+            
```
Revert this file? #Closed

MGudgin commented on targets/csharp/templates/PlayFabAuthenticationContext.cs.ejs:
```diff
@@ -4,15 +4,15 @@ namespace PlayFab
     {
```
Why did we need the changes in this file? #Closed

nima-ap-zz replied:
Repeating what I told you in person here for the sake of logging it.
`PlayFabCommonRequest` now has this field and hence every `_______Request` does too. Since we're trying to now pass the authentication information in the request here instead of the header, we need a property with `{ get; set; }` instead of a field due to JSON serialization/deserialization concerns. #Closed

MGudgin commented on targets/csharp/templates/PlayFabSettings.cs.ejs:
```diff
@@ -24,6 +24,25 @@ namespace PlayFab
+		/// <summary> Placeholder for the Local Debugging URI. </summary>
+		private static string _debugUri;
+		/// <summary> Must be manually set by user on .NETStandard 1.1 and .NETPortable for local debugging to work properly </summary>
+		public static string DebugUri 
```
This property doesn't look right. You should suppress the setter in the none NET45/NETCOREAPP2_0 case otherwise the behaviour is weird.

Also the tabbing looks to be indented too far from line 27 to 44 #Closed

nima-ap-zz replied:
In that case i'll wrap the whole setter method with `if net45 || net2_0` so that it turns into a read only property in those cases. #Closed

MGudgin commented on targets/csharp/templates/PlayFabSettings.cs.ejs:
```diff
@@ -44,13 +63,14 @@ namespace PlayFab
             var baseUrl = apiSettings?.ProductionEnvironmentUrl ?? ProductionEnvironmentUrl;
             if (!baseUrl.StartsWith("http"))
             {
+				sb.Append("https://");
```
Tabbing looks wrong here too. #Closed

MGudgin commented on targets/csharp/templates/PlayFabUtil.cs.ejs:
```diff
@@ -81,5 +83,79 @@ namespace PlayFab
+        /// Load a text file, and return the file as text.
+        /// Used for small (usually json) files.
+        /// </summary>
+        public static string ReadAllFileText(string filename)
```
Why is this public? #Closed

nima-ap-zz replied:
It was there before our current work but since there are no more references to it besides these changes I will make it private. #Closed

MGudgin commented on targets/csharp/templates/PlayFabUtil.cs.ejs:
```diff
@@ -81,5 +83,79 @@ namespace PlayFab
+        /// </summary>
+        public static string ReadAllFileText(string filename)
+        {
+#if NET45 || NETCOREAPP2_0
```
Why not just put the #if around the entire function? We don't call the function on other platforms, right? You could also put an #if around the ThreadStatic too then. #Closed

MGudgin commented on targets/csharp/templates/PlayFabUtil.cs.ejs:
```diff
@@ -81,5 +83,79 @@ namespace PlayFab
+
+        internal static string GetLocalSettingsFileProperty(string propertyKey)
+        {
+#if NET45 || NETCOREAPP2_0
```
Same comment here about having the function just not show up outside NET45/NETCOREAPP2_0 #Closed

nima-ap-zz replied:
Moved the #if to become 1 wrap around both `ReadAllFileText` and `GetLocalSettingsFileProperty` #Closed

MGudgin commented on targets/csharp/templates/PlayFabUtil.cs.ejs:
```diff
@@ -81,5 +83,79 @@ namespace PlayFab
             return Sb.ToString();
         }
+
+				[ThreadStatic]
```
Tabbing #Closed

MGudgin commented on targets/unity-v2/make.js:
```diff
@@ -379,8 +379,12 @@ function getAuthParams(apiCall) {
 }
 
 function getRequestActions(tabbing, apiCall) {
```
As above, I think we could be emitting incorrect code in the LoginResult/RegisterPlayFabUserRequest case. Let's just use return statements and have each if block contain all the code needed for that case. #Closed

MGudgin commented on targets/unity-v2/source/Shared/Internal/PlayFabHttp/PlayFabHTTP.cs:
```diff
@@ -179,12 +179,30 @@ public static void SimplePostCall(string fullUrl, byte[] payload, Action<byte[]>
+            _MakeApiCall(apiEndpoint, fullUrl, request, authType, resultCallback, errorCallback, customData, extraHeaders, allowQueueing);
+        }
+
+        protected internal static void DebugApiCall<TResult>(string fullUri,
```
Let's call this MakeDebugApiCall or MakeApiCallWithFullUri #Closed

nima-ap-zz replied:
`MakeApiCallWithFullUri` it is 👍  #Closed

MGudgin commented on targets/unity-v2/source/Shared/Internal/Util.cs:
```diff
@@ -74,16 +74,66 @@ public static string Format(string text, params object[] args)
+                fs = new FileStream(filename, FileMode.Open);
+                using (BinaryReader br = new BinaryReader(fs))
+                {
+                    // Attempt to release the descriptor immediately
```
Why? Don't you need a using for the FileStream anyway to ensure it get's disposed? #Closed

MGudgin commented on targets/unity-v2/source/Shared/Internal/Util.cs:
```diff
@@ -74,16 +74,66 @@ public static string Format(string text, params object[] args)
         /// </summary>
         public static string ReadAllFileText(string filename)
         {
```
Please use braces even for single line blocks #Closed

MGudgin commented on targets/unity-v2/source/Shared/Internal/Util.cs:
```diff
@@ -74,16 +74,66 @@ public static string Format(string text, params object[] args)
+            if (!File.Exists(filename))
+                return string.Empty;
+                
             if (_sb == null)
```
is _sb a ThreadStatic here too (I can't see that part of the file here) #Closed

nima-ap-zz replied:
Yep it's on line 68. #Closed

MGudgin commented on targets/unity-v2/source/Shared/Internal/Util.cs:
```diff
@@ -74,16 +74,66 @@ public static string Format(string text, params object[] args)
+            }
+            finally
+            {
+                // Make sure to dispose of the file stream in case the construction of binary reader fails
```
Ah, I see the dispose here. If you just did a using for fs though, you wouldn't need the finally block, would you? #Closed

nima-ap-zz replied:
We do `using` on `using(BinaryReader br = new BinaryReader(fs))` and in case that instantiation fails we want to make sure we still get rid of the file descriptor. #Closed

nima-ap-zz replied:
Let's do `using` on both disposables actually. #Closed

MGudgin commented on targets/unity-v2/source/Shared/Internal/Util.cs:
```diff
@@ -74,16 +74,66 @@ public static string Format(string text, params object[] args)
+        {
+            string envFileContent = null;
+            
+            string envFileName = "playfab.local.settings.json";
```
Should we move this out to be a private field like in the C# SDK? #Closed

nima-ap-zz replied:
Done #Closed

MGudgin commented on targets/unity-v2/source/Shared/Public/PlayFabSettings.cs.ejs:
```diff
@@ -159,13 +159,14 @@ namespace PlayFab
             var baseUrl = ProductionEnvironmentUrlPrivate;
             if (!baseUrl.StartsWith("http"))
             {
+				sb.Append("https://");
```
Tabbing #Closed

nima-ap committed:
Address Gudge comments on PR

1. Fix tabbing in `.ejs` files to spaces instead of tabs to match the rest of the project.
2. Revert `getRequestActions` in `make.js` so that it now has `return` statements in all cases instead of appending to an `output` variable.
3. Extract the custom behavior of `ExecuteFunction` API call into its own function in `make.js`: `getCustomApiLogic`
4.  Use another `using` for `FileStream` in `Util.cs` for Unity SDK to avoid the `try, finally` block as a way of disposing it in case of failure.

nima-ap commented:
All comments so far have been addressed.

nima-ap committed:
Fix tabbing in PlayFabSettings.cs.ejs

nima-ap committed:
Rename Unity `DebugApiCall` to `MakeApiCallWithFullUri`

MGudgin reviewed (approved):
:shipit:
